duqtools/operations.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+import logging
+from collections import deque
+from inspect import signature
+from typing import Callable
+
+import click
+from pydantic import Field
+
+from .schema import BaseModel
+
+logger = logging.getLogger(__name__)
+
+
+class Operation(BaseModel):
+    """Operation, simple class which has a callable action."""
+
+    description: str = Field(
+        description='description of the operation to be done')
+    action: Callable = Field(
+        description='a function which can be executed when we '
+        'decide to apply this operation')
+    args: tuple = Field((),
+                        description='positional arguments that have to be '
+                        'passed to the action')
+    kwargs: dict = Field({},
+                         description='keyword arguments that will be '
+                         'passed to the action')
+
+    def __call__(self) -> Operation:
+        logger.info(f'-: {self.description}')
+        self.action(*self.args, **self.kwargs)
+        return self
+
+
+class Operations(deque):
+    """Operations Queue which keeps track of all the operations that need to be
+    done."""
+
+    _instance = None
+    yes = False  # Apply operations without prompt
+    dry_run = False  # Never apply any operations (do not even ask)
+
+    def __new__(cls, *args, **kwargs):
+        # Make it a singleton
+        if not Operations._instance:
+            Operations._instance = super().__new__(cls)
+        return Operations._instance
+
+    def add(self, *args, **kwargs) -> None:
+        """convenience Operation wrapper around put."""
+
+        self.append(Operation(*args, **kwargs))
+
+    def put(self, item: Operation) -> None:
+        """synonym for append."""
+        self.append(item)
+
+    def append(self, item: Operation) -> None:  # type: ignore
+        """Restrict our diet to Operation objects only."""
+
+        super().append(item)
+
+    def apply(self) -> Operation:
+        """Apply the next operation in the queue and remove it."""
+
+        return self.popleft()()
```

Comment:
It's not entirely clear to me why this should return the operation. 

It is more clear to me what this is doing if you pop the next action first, and then call it.

```suggestion
        next_op = self.popleft()
        next_op()
```

Reply:
You got me, I found it too wonderfully obfuscating that I left it in, but I'll change it ;)